### ISSUE TITLE
key-value cannot be used as arguments to other job in workflow

### ISSUE THREAD

**Issue by krzysbaranski** [rundeck/rundeck#2752]
Issue type: Bug report

**My Rundeck detail**

* Rundeck version: Rundeck 2.9.3-1 
* install type: vagrant/anvils-demo; rpm
* OS Name/version: centos ; debian
* DB Type/version: h2

**Expected Behavior**
- when passing key-value variable to step *job reference* via argument - value should be replaced by captured key-value data
example argument in job reference: `-referenceTime ${data.reftime}`

**Actual Behavior**
data is not replaced and it passes `${data.reftime}` data as is without replacement

Similar error for *inline script* with inline replacement `echo "@data.reftime@"`
when using first node strategy in job output is empty,
but works only for sequential strategy

**How to reproduce Behavior**
import jobs and run:
https://gist.github.com/krzysbaranski/ee1c05cde416b2627fb004430ea9a87a
https://gist.github.com/krzysbaranski/39bb9d9f776012d30b97e74ce95c58db

docs: #2577

**Comment by gschueler** [comment]
try 2.9.4. when in node-first strategy, some data was not getting passed correctly, that was fixed in 2.9.4

**Comment by krzysbaranski** [comment]
It's working in 2.10.1-1 - Thanks.

I updated job in gist with `nodeStep: 'true'`

**Comment by vvbogdanov87** [comment]
The bug exists in rundeck v3.0.2.
```
RUNDECK:DATA:timestamp=1535029513421

Key Value Data: Results
--
timestamp | 1535029513421

/tmp/203-54-localhost-dispatch-script.tmp.sh: line 6: ${data.timestamp}: bad substitution
```

**Comment by gschueler** [comment]
@vvbogdanov87 what is your job definition?

**Comment by vvbogdanov87** [comment]
```yaml
- defaultTab: summary
  description: hidden
  executionEnabled: true
  id: hidden
  loglevel: INFO
  loglimit: 100MB
  loglimitAction: halt
  loglimitStatus: failed
  multipleExecutions: true
  name: hidden
  nodeFilterEditable: true
  notification:
    onfailure:
      email:
        recipients: hidden
        subject:hidden
  notifyAvgDurationThreshold: null
  options:
  - delimiter: ','
    description: hidden
    multivalued: true
    name: feed
  - name: memory
    regex: \d+[MG]
    value: 2G
  scheduleEnabled: true
  sequence:
    commands:
    - plugins:
        LogFilter: []
      script: |
        #!/bin/bash
        timestamp=`date +%s%3N`
        echo "RUNDECK:DATA:timestamp=${timestamp}"

        feed=@option.feed@
        feed=${feed//,/-}
        feed=${feed//_/}
        echo "RUNDECK:DATA:jname=${feed}"
    - script: |-
        #!/bin/bash
        echo @data.timestamp@
        echo @data.jname@
    - description: hidden
      jobref:
        args: -name ${data.jname} -memory ${option.memory} -options ${data.timestamp}
        group: ''
        name: hidden
        uuid: hidden
    keepgoing: false
    pluginConfig:
      LogFilter:
      - config:
          logData: 'true'
          regex: ^RUNDECK:DATA:(.+?)\s*=\s*(.+)$
        type: key-value-data
    strategy: sequential
  uuid: hidden
```

**Comment by vvbogdanov87** [comment]
Checked on rundeck 2.10.1-1. Result is the same.

**Comment by vvbogdanov87** [comment]
Sorry. It works on both versions if `nodeStep: 'true'`

**Comment by gschueler** [comment]
if you try to reference data capture in node context (e.g script/command output on a node) from a step that is not a node context, the data is not available via `${data.timestamp}`.  You can collate all of the node values using `${data.timestamp*}`.  If you are only running on a single node, the value will be the single value, otherwise it will be comma-separated collection of all the node values.

**Comment by krzysbaranski** [comment]
thanks for a tip @gschueler  I was looking for this feature `${data.timestamp*}`

This feature was released in 2.10.1 https://rundeck.org/docs/history/changelog.html#release-2.10.1
